File: src/main/java/dev/nytweetdeck/timeline/TimelineQueryFactory.java
package dev.nytweetdeck.timeline;

import java.util.LinkedHashMap;
import java.util.Map;
import org.springframework.stereotype.Component;

@Component
public class TimelineQueryFactory {

    public Query create(String kind, String target, String cursor) {
        return create(kind, target, cursor, "latest");
    }

    public Query create(String kind, String target, String cursor, String sort) {
        var normalizedSort = normalizeSort(sort);
        var variables = new LinkedHashMap<String, Object>();
        variables.put("count", 20);
        if (cursor != null && !cursor.isBlank()) {
            variables.put("cursor", cursor);
        }
        return switch (kind) {
            case "homeForYou" -> {
                variables.put("enableRanking", "top".equals(normalizedSort));
                variables.put("includePromotedContent", false);
                variables.put("requestContext", "launch");
                variables.put("withCommunity", true);
                yield new Query("homeForYou", variables);
            }
            case "homeFollowing" -> {
                variables.put("enableRanking", "top".equals(normalizedSort));
                variables.put("includePromotedContent", false);
                variables.put("requestContext", "launch");
                yield new Query("homeFollowing", variables);
            }
            case "userPosts" -> {
                variables.put("userId", requireTarget(target, kind));
                variables.put("includePromotedContent", false);
                variables.put("withQuickPromoteEligibilityTweetFields", false);
                variables.put("withVoice", true);
                yield new Query("userPosts", variables);
            }
            case "list" -> {
                variables.put("listId", requireTarget(target, kind));
                yield new Query("list", variables);
            }
            case "history" -> new Query("history", variables);
            case "trends" -> new Query("trends", variables);
            case "notifications" -> new Query("notifications", variables);
            case "search" -> {
                variables.put("rawQuery", requireTarget(target, kind));
                variables.put("querySource", "typed_query");
                variables.put("product", "top".equals(normalizedSort) ? "Top" : "Latest");
                variables.put("withGrokTranslatedBio", false);
                variables.put("withQuickPromoteEligibilityTweetFields", false);
                yield new Query("search", variables);
            }
            default -> throw new IllegalArgumentException("未対応のタイムライン種別です: " + kind);
        };
    }

    private static String normalizeSort(String sort) {
        var normalized = sort == null ? "latest" : sort.trim().toLowerCase(java.util.Locale.ROOT);
        if (!normalized.equals("latest") && !normalized.equals("top")) {
            throw new IllegalArgumentException("カラムの並び順が不正です。");
        }
        return normalized;
    }

    private static String requireTarget(String target, String kind) {
        if (target == null || target.isBlank() || target.length() > 200) {
            throw new IllegalArgumentException(kind + "には有効な対象が必要です。");
        }
        return target;
    }

    public record Query(String purpose, Map<String, Object> variables) {
        public Query {
            variables = Map.copyOf(variables);
        }
    }
}

File: android/app/src/main/java/dev/nytweetdeck/android/xapi/TimelineQueryFactory.kt
package dev.nytweetdeck.android.xapi

import java.util.Locale

data class TimelineQuery(
    val purpose: String,
    val variables: Map<String, Any?>,
)

object TimelineQueryFactory {
    fun create(kind: String, target: String?, cursor: String?, sort: String = "latest"): TimelineQuery {
        val normalizedSort = sort.trim().lowercase(Locale.ROOT).also {
            require(it == "latest" || it == "top") { "カラムの並び順が不正です。" }
        }
        val variables = linkedMapOf<String, Any?>("count" to 20)
        cursor?.takeIf(String::isNotBlank)?.let { variables["cursor"] = it }
        val purpose = when (kind) {
            "homeForYou" -> {
                variables["enableRanking"] = normalizedSort == "top"
                variables["includePromotedContent"] = false
                variables["requestContext"] = "launch"
                variables["withCommunity"] = true
                "homeForYou"
            }
            "homeFollowing" -> {
                variables["enableRanking"] = normalizedSort == "top"
                variables["includePromotedContent"] = false
                variables["requestContext"] = "launch"
                "homeFollowing"
            }
            "userPosts" -> {
                variables["userId"] = requireTarget(target, kind)
                variables["includePromotedContent"] = false
                variables["withQuickPromoteEligibilityTweetFields"] = false
                variables["withVoice"] = true
                "userPosts"
            }
            "list" -> {
                variables["listId"] = requireTarget(target, kind)
                "list"
            }
            "history", "trends", "notifications" -> kind
            "search" -> {
                variables["rawQuery"] = requireTarget(target, kind)
                variables["querySource"] = "typed_query"
                variables["product"] = if (normalizedSort == "top") "Top" else "Latest"
                variables["withGrokTranslatedBio"] = false
                variables["withQuickPromoteEligibilityTweetFields"] = false
                "search"
            }
            else -> throw IllegalArgumentException("未対応のタイムライン種別です: $kind")
        }
        return TimelineQuery(purpose, variables.toMap())
    }

    private fun requireTarget(value: String?, kind: String): String {
        require(value != null && value.isNotBlank() && value.length <= 200) {
            "${kind}には有効な対象が必要です。"
        }
        return value
    }
}
